atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasBackupService.java

```diff
@@ -113,46 +116,70 @@ public static AtlasBackupService create(
     }
 
     public static AtlasBackupService createForTests(
             AuthHeader authHeader,
             AtlasBackupClient atlasBackupClient,
             TransactionManager transactionManager,
             Function<AtlasService, Path> backupFolderFactory) {
+        return createAtlasBackupService(
+                authHeader, atlasBackupClient, backupFolderFactory, KvsRunner.create(transactionManager), 10);
+    }
+
+    private static AtlasBackupService createAtlasBackupService(
+            AuthHeader authHeader,
+            AtlasBackupClient atlasBackupClient,
+            Function<AtlasService, Path> backupFolderFactory,
+            KvsRunner kvsRunner,
+            int completeBackupNumThreads) {
         BackupPersister backupPersister = new ExternalBackupPersister(backupFolderFactory);
-        KvsRunner kvsRunner = KvsRunner.create(transactionManager);
         CoordinationServiceRecorder coordinationServiceRecorder =
                 new CoordinationServiceRecorder(kvsRunner, backupPersister);
-        LockRefresher<InProgressBackupToken> lockRefresher = getLockRefresher(authHeader, atlasBackupClient);
+        ScheduledExecutorService refreshExecutor =
+                PTExecutors.newSingleThreadScheduledExecutor(new NamedThreadFactory("backupLockRefresher", true));
+        LockRefresher<InProgressBackupToken> lockRefresher =
+                getLockRefresher(authHeader, atlasBackupClient, refreshExecutor);
 
         return new AtlasBackupService(
-                authHeader, atlasBackupClient, coordinationServiceRecorder, backupPersister, lockRefresher, 10);
+                authHeader,
+                atlasBackupClient,
+                coordinationServiceRecorder,
+                backupPersister,
+                lockRefresher,
+                refreshExecutor,
+                completeBackupNumThreads);
     }
 
     private static LockRefresher<InProgressBackupToken> getLockRefresher(
-            AuthHeader authHeader, AtlasBackupClient atlasBackupClient) {
-        ScheduledExecutorService refreshExecutor =
-                PTExecutors.newSingleThreadScheduledExecutor(new NamedThreadFactory("backupLockRefresher", true));
+            AuthHeader authHeader, AtlasBackupClient atlasBackupClient, ScheduledExecutorService refreshExecutor) {
         LockLeaseRefresher<InProgressBackupToken> lockLeaseRefresher = tokens -> atlasBackupClient
                 .refreshBackup(authHeader, RefreshBackupRequest.of(tokens))
                 .getRefreshedTokens();
         return new LockRefresher<>(refreshExecutor, lockLeaseRefresher, REFRESH_INTERVAL_MILLIS);
     }
 
+    /**
+     * Cleans up the backup operation. This will release all immutable timestamp locks.
+     * Must be called after any failure; after calling this, completeBackup will fail for any namespace which
+     * has not yet reached that stage.
+     */
+    public void cleanupBackup() {
```

Comment:
What happens if you try and prepare a backup at this point? I think you probably also want this to set a boolean flag (i.e. `isOpen` or something), then either throw or fail or whatever is standard on all of your public methods if they are called and you are no longer open, as it looks like once we've actually _closed_ the lock refresher and executor, we really don't want to be doing anything more here.

Also, nit: maybe this _should_ be called close: we're not really "cleaning up" a backup, but rather closing the backup service itself.

Reply:
Yeah, this all makes sense. I'll add an `AtomicBoolean isClosed`, and we can throw exceptions on `prepare` or `complete` if called after `close`.

Reply:
a volatile is technically sufficient, but sure (:

Reply:
done (with volatility)